docs/prover/approx/grounding.md

```diff
@@ -0,0 +1,274 @@
+require forall mathint x . x * 2 != y;
+require forall uint x . forall mathint y . f(x) != 2 * y;
+```
+
+In the first example, `x` is not used as an argument to a function, while in the
+second case, `y` is not.
+
+
+Although you are allowed to call functions on complicated expressions that use
+quantified variables, doing so may produce spurious counterexamples.  For
+example, the following is allowed, but is likely to produce spurious
+counterexamples (because `x` itself is not an argument to a function):
+
+```{todo}
+I'm sort of surprised that we allow this.  It would be simpler if we just required
+that `x` must be used as a simple argument to some function, right?
+```
+
+```cvl
+require forall uint x . f(2 * x) == 0;
+```
+
+This particular example requires that all even inputs to `f` produce `0` as
+output; it could be rewritten as follows:
+
+```cvl
+require forall uint y . (y % 2 == 0) => f(y);
+```
+
+If you use a quantified variable in an argument to two different functions,
+you may produce spurious counterexamples.  For example:
+
+```cvl
+require forall uint x . f(x) <= g(x+1) && g(x+1) != 0;
+```
+
+Here, `x + 1` is used as an argument to `g`, but `x` is not; you may get
+counterexamples where `g(x+1) == 0` for some `x`.  In that case, you can add an
+additional equivalent require that does use the quantified variable as an argument
+to `g`:
+
+```cvl
+require forall uint x . f(x) < g(x+1) && g(x+1) != 0;
+require forall uint x . f(x-1) < g(x) && g(x) != 0;
+```
+
+This will make `x` a direct argument to `g`, so the expression will be
+grounded properly.
+
+(grounding-polarity)=
+### Double Polarity
+
+The polarity of a sub-formula is the direction of the effect it has on the
+output of the formula. This is best demonstrated through an example:
+
+```cvl
+a && (b || !c)
+```
+
+In this example, `b` possesses a positive polarity because changing `b` from `false` to `true`
+can't make the formula `false` if it wasn't before.  Informally, making `b`
+"more true" can only make the whole formula "more true".
+
+In this example, `a` also possesses positive polarity: if the formula was true when `a` was
+`false`, it must also be `true` when `a` is `true`.
+
+On the other hand, `c` has a negative polarity because changing `c` from `true`
+to `false` can only make the statement "more false".  If can't become `true` if
+it wasn't true before.
+
+Sub-expressions can also have double polarity.  For example, consider the
+formula
+```
+a <=> b
+```
+In this example, `a` has double polarity, because making `a` true could cause
+the formula to become `false` when it was true before, but it could also cause
+the formula to become `true` when it wasn't before.
+
+Grounding is disallowed when the quantified expression has double polarity in the
+rule.  For example, the following is disallowed, because the `forall` statement
+has double polarity.
+```cvl
+assert (forall uint x . f(x) > 0) <=> y;
+```
+
+```{todo}
+advice for rewriting?
+```
+
```

Comment:
Split to two asserts, in each one the quantified expression has a single polarity:
```
assert (forall uint x . f(x) > 0) => y;
assert y => (forall uint x . f(x) > 0);
```